Add syncPublicationOnboardingState action

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8796 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.test.js:
```diff
@@ -16,115 +17,202 @@
 			} );
 
-			it( 'should not make a network request if publications are already present', async () => {
+			it( 'should update the settings and call the saveSettings endpoint', async () => {
```
I think this would be an appropriate test to mock `Date.now()` in, so as to provide a more precise assertions below:

```suggestion
			it( 'should update the settings and call the saveSettings endpoint', async () => {
				const originalDateNow = Date.now;
				// Mock the date to be an arbitrary time.
				const mockNow = new Date( '2020-01-01 12:30:00' ).getTime();
				Date.now = jest.fn( () => mockNow );
```

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.test.js:
```diff
@@ -16,115 +17,202 @@
+						...settings,
+						publicationOnboardingState:
+							PUBLICATION_ONBOARDING_STATES.UNSPECIFIED,
+						publicationOnboardingStateLastSyncedAtMs: Date.now(),
```
If we're using a mocked `Date.now()` value, this can be updated to reflect it:

```suggestion
						publicationOnboardingStateLastSyncedAtMs: mockNow,
```

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.test.js:
```diff
@@ -16,115 +17,202 @@
-					fixtures.publications.length
-				);
+				// Ensure that the sync time is set.
+				expect( syncTimeMs ).not.toBe( 0 );
+
+				// Ensure that date is within the last 5 seconds.
+				expect( syncTimeMs ).toBeLessThanOrEqual( Date.now() );
+
+				expect( syncTimeMs ).toBeGreaterThan( Date.now() - 5000 );
```
If we're using the mocked `Date.now()` this can be simplified a little and made more precise:

```suggestion
				expect( syncTimeMs ).toBe( mockNow );

				// Reset the date.
				Date.now = originalDateNow;
```

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.test.js:
```diff
@@ -16,115 +17,202 @@
+					publicationOnboardingStateLastSyncedAtMs: 0,
+				};
+
+				fetchMock.post( settingsEndpoint, {
```
We tend to use `postOnce()` over `post()` where we can:

```suggestion
				fetchMock.postOnce( settingsEndpoint, {
```

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.test.js:
```diff
@@ -16,115 +17,202 @@
-				fetchMock.getOnce( publicationsEndpoint, {
-					body: response,
-					status: 500,
+				fetchMock.post( settingsEndpoint, {
```
As above:

```suggestion
				fetchMock.postOnce( settingsEndpoint, {
```

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.test.js:
```diff
@@ -16,115 +17,202 @@
-					.getPublications();
-				expect( publications ).toBeUndefined();
-				expect( console ).toHaveErrored();
+				const uiKeyBeforeAction = registry
+					.select( CORE_UI )
+					.getValue(
+						UI_KEY_SHOW_RRM_PUBLICATION_APPROVED_NOTIFICATION
+					);
```
We might as well assert here - it means the test will fail sooner rather than potentially get stuck in an await and not log a failed assertion for this value; it would also be more consistent with other tests:

```suggestion
				expect(
					registry
						.select( CORE_UI )
						.getValue(
							UI_KEY_SHOW_RRM_PUBLICATION_APPROVED_NOTIFICATION
						)
				).toBe( undefined );
```

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.test.js:
```diff
@@ -16,115 +17,202 @@
+					.dispatch( MODULES_READER_REVENUE_MANAGER )
+					.syncPublicationOnboardingState();
+
+				const uiKeyAfterAction = registry
+					.select( CORE_UI )
+					.getValue(
+						UI_KEY_SHOW_RRM_PUBLICATION_APPROVED_NOTIFICATION
+					);
+
+				// Ensure that the UI key is set to true.
+				expect( uiKeyBeforeAction ).toBe( undefined );
+				expect( uiKeyAfterAction ).toBe( true );
```
Following on from the comment above:

```suggestion
				// Ensure that the UI key is set to true.
				expect(
					registry
						.select( CORE_UI )
						.getValue(
							UI_KEY_SHOW_RRM_PUBLICATION_APPROVED_NOTIFICATION
						)
				).toBe( true );
```

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.js:
```diff
@@ -45,6 +49,94 @@ const baseInitialState = {
+			// eslint-disable-next-line sitekit/acronym-case
+			( { publicationId } ) => publicationId === publicationID
+		);
+
```
We should also add a guard here in case the publication is not found:

```suggestion

		// If the publications is not found, do not attempt to sync the onboarding state.
		if ( ! publication ) {
			return;
		}
```

techanvil reviewed (changes requested):
Hey @ankitrox, nice work here so far. I have left a few comments, please take a look.

ankitrox commented:
Thank you @techanvil for reviewing the PR and adding your valuable feedback. I would also like to thank you for your guidance on testing `Date.now()`, it was an absolute help and to the point!

I have addressed your feedback and sending your way for another review.

Thanks again.

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.js:
```diff
@@ -89,32 +89,36 @@ const baseActions = {
 			( { publicationId } ) => publicationId === publicationID
 		);
 
+		// If the publications is not found, do not attempt to sync the onboarding state.
```
Fix typo:

```suggestion
		// If the publication is not found, do not attempt to sync the onboarding state.
```

techanvil commented on assets/js/modules/reader-revenue-manager/datastore/publications.js:
```diff
@@ -45,6 +49,98 @@ const baseInitialState = {
 const baseActions = {
+	/**
+	 * Syncronizes the onboarding state of the publication with the API.
+	 * Updates the settings on server.
```
Fix typo.

```suggestion
	 * Updates the settings on the server.
```

techanvil reviewed (approved):
LGTM, nice one @ankitrox!

aaemnnosttv commented on assets/js/modules/reader-revenue-manager/datastore/publications.js:
```diff
@@ -17,22 +17,26 @@
 import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
 import {
 	MODULES_READER_REVENUE_MANAGER,
+	MODULE_SLUG,
```
We have an (old) related issue about this. See #3417.

It's worth noting that we included the name in the constant for module stores which previously all used `STORE_NAME` internally and aliases externally. So this kind of generic constant is one we'll want to avoid here as well.

@nfmohit @techanvil

ankitrox replied:
Thanks for flagging this @aaemnnosttv .

I will rename this to `READER_REVENUE_MANAGER_MODULE_SLUG` in subsequent issue #8846

aaemnnosttv replied:
Let's keep the `SLUG` part as a prefix rather than a suffix, similar to the store name constant.

aaemnnosttv replied:
Honestly, I'd be happy to use `RRM` everywhere instead of the full name which is quite long as can be seen here 😅

techanvil replied:
I should mention here that @ankitrox had previously used `RRM` in another constant name but [I advised him against it](https://github.com/google/site-kit-wp/pull/9017#discussion_r1682568311), referring to our naming conventions whereby we avoid abbreviations in variable names etc.

Tbh, I was wondering if I was being a bit dogmatic, but was simply going on memories of past feedback. I'm entirely happy if we take a more pragmatic approach and use the RRM abbreviation for names in the codebase!

aaemnnosttv commented on assets/js/modules/reader-revenue-manager/datastore/publications.js:
```diff
@@ -17,22 +17,26 @@
+			MODULE_SLUG,
 			'publications',
 			{},
 			{ useCache: true }
```
Unrelated observation – but selectors like this one which are only really used during setup should not use cache so that it pulls fresh data every time, otherwise there is no way to pull a new publication after creating one on the service once cached.

techanvil replied:
Good spot @aaemnnosttv.

It would be nice to roll this one-line fix into a forthcoming RRM PR rather than create a separate issue for it.

@ankitrox, could I ask you to please include it in your next PR?

```suggestion
			{ useCache: false }
```

nfmohit replied:
Thank you folks! Just to make sure we don't miss this out, I've included this in [my PR](https://github.com/google/site-kit-wp/pull/9098) for #8800.

aaemnnosttv commented on assets/js/modules/reader-revenue-manager/datastore/publications.js:
```diff
@@ -94,6 +190,7 @@ const baseResolvers = {
 			.getPublications();
 		if ( publications === undefined ) {
 			yield fetchGetPublicationsStore.actions.fetchGetPublications();
+			yield baseActions.syncPublicationOnboardingState();
```
@techanvil @ankitrox Why did we add this here? Syncing the onboarding state only makes sense in the context of a selected publication in settings which this selector/resolver is not related to. Also, since this will usually be called before any selection is made I would expect this to not have the intended result. We should only sync the onboarding state for the publication which is saved in settings, not simply selected in state.

techanvil replied:
@aaemnnosttv this is defined as a requirement in the [design doc](https://docs.google.com/document/d/14CxsQjTpdPrQpd8NoQtXnUd1DG7FVtmF5GTVvYdixVY/edit#bookmark=id.9bqgtkjartaf), and the AC/IB for this issue. So, I haven't really questioned it. My presumption is that the intention is to resync the onboarding state for the currently selected publication any time we visit the Settings screen after it has been connected. Arguably, though, it could make sense to decouple the sync from `getPublications()`. Maybe @nfmohit can shed a bit more light on this one?

nfmohit replied:
Thank you for raising this, @aaemnnosttv. @techanvil is correct with the presumption above. Since we show the `PublicationOnboardingStateNotice` in the settings screen based on the `publicationOnboardingState` module setting which is updated once every hour, there is a chance we may show an unintended notice due to the onboarding state not being up to date.

Since the `getPublications` selector would likely be resolved in `SettingsEdit`, my intention was to dispatch the `syncPublicationOnboardingState` action that would update the onboarding state in the module settings. The action only does this if the module is connected and a publication ID is set.

This would ensure that the correct `PublicationOnboardingStateNotice` is shown (or not shown at all). Would you suggest we do this any differently?

nfmohit replied:
Now that I think deeply about this, I feel this is not really necessary and does not add much value and has the potential to cause problems as Evan mentioned above. @techanvil What do you think of the idea of removing this dispatch as part of a subsequent issue?

aaemnnosttv replied:
If the intention is to sync the state on the settings view, then I think it would be more appropriate to move the effect to that component as it's not related to this selector.

techanvil replied:
Thanks @nfmohit and @aaemnnosttv. Agreed, it would be good to move the sync out to the component.

I think this merits a new issue to flesh out the details. We also need to revisit the `PublicationSelect` component due to the divergence from the IB in PR https://github.com/google/site-kit-wp/pull/9006:

> Note: `publicationOnboardingStateLastSyncedAtMs` is not getting updated as part of `onPublicationChange` function because it will be updated as part of the `syncPublicationOnboardingState` action, which will run immediately before (`getPublications` resolver)/after (`maybeSyncPublicationsOnboardingState`) a publication is selected.

@nfmohit, would you be happy to create an issue for this?

nfmohit replied:
> Agreed, it would be good to move the sync out to the component.
> 
> I think this merits a new issue to flesh out the details.

Thank you @techanvil. About this one, now that I think about it, I feel updating it in `SettingsEdit` might actually have unexpected results. The action is responsible for showing the publication approved overlay notification by changing a `core/ui` data store key, which would only happen when the action is dispatched in the main dashboard. Also, the state would be updated once every hour anyway, so just updating it only for `SettingsEdit` doesn't seem to be worth it while the publication onboarding state notice also appears in `SettingsView`.

> I think this merits a new issue to flesh out the details. We also need to revisit the `PublicationSelect` component due to the divergence from the IB in PR #9006:
> 
> > Note: `publicationOnboardingStateLastSyncedAtMs` is not getting updated as part of `onPublicationChange` function because it will be updated as part of the `syncPublicationOnboardingState` action, which will run immediately before (`getPublications` resolver)/after (`maybeSyncPublicationsOnboardingState`) a publication is selected.

About this one, I also feel updating `publicationOnboardingStateLastSyncedAtMs` as part of `syncPublicationOnboardingState` is sufficient.

Please let me know what you think, thanks so much!

techanvil replied:
> Thank you @techanvil. About this one, now that I think about it, I feel updating it in `SettingsEdit` might actually have unexpected results. The action is responsible for showing the publication approved overlay notification by changing a `core/ui` data store key, which would only happen when the action is dispatched in the main dashboard. Also, the state would be updated once every hour anyway, so just updating it only for `SettingsEdit` doesn't seem to be worth it while the publication onboarding state notice also appears in `SettingsView`. To be fair, dispatching this as a part of the resolver was planned for redundancy.

Thanks @nfmohit. As you've pointed out, the publication onboarding state notice appears both in `SettingsEdit` and `SettingsView`, so it's a bit of an oddity to only sync the state in `SettingsEdit`. I think we could either dispatch the sync in `SettingsView` instead, ensuring the notice is up-to-date in both places, or not at all, and allow for the fact it could be stale. Obviously from UX perspective it would be better if it was up-to-date, so I'd lean in that direction; that said we might then want to consider a loading state which could add a bit of complexity. Maybe we can simply omit it for now and add it as an improvement post launch.

If we do still sync on the Settings screen, setting the `core/ui` value won't have any effect so I don't see it being too much of a concern here; that said we could still amend the action so as to avoid that if not on the dashboard.

> > > Note: `publicationOnboardingStateLastSyncedAtMs` is not getting updated as part of `onPublicationChange` function because it will be updated as part of the `syncPublicationOnboardingState` action, which will run immediately before (`getPublications` resolver)/after (`maybeSyncPublicationsOnboardingState`) a publication is selected.
> 
> About this one, I also feel updating `publicationOnboardingStateLastSyncedAtMs` as part of `syncPublicationOnboardingState` is sufficient.

Hmm. If we are removing the call to `syncPublicationOnboardingState()` from the Settings screen entirely, then the logic about it being run immediately before a publication is selected no longer holds true.

So, we could find ourselves making another call to the API in order to sync the onboarding state when we land on the dashboard, even though we've just updated the value.

I would say it would be preferable to avoid this to help reduce unneeded API requests and the relevant API quota usage. So my preference would be to update the timestamp if we are not syncing on the Settings screen. However, it's a pretty marginal point and again something we could address post launch.

nfmohit replied:
Thank you for your input, Tom! I've included updating this module setting as part of my `selectPublication` action proposed in [my PR](https://github.com/google/site-kit-wp/pull/9098) for [#8800](https://github.com/google/site-kit-wp/issues/8800).

techanvil replied:
Looks good, nice one Nahid!

nfmohit commented on assets/js/modules/reader-revenue-manager/datastore/publications.js:
```diff
@@ -45,6 +49,98 @@ const baseInitialState = {
+		// Save the settings to the API.
+		registry.dispatch( MODULES_READER_REVENUE_MANAGER ).saveSettings();
+
+		// If the onboarding state is complete, set the key in CORE_UI to trigger the notification.
+		if (
+			onboardingState ===
+			PUBLICATION_ONBOARDING_STATES.ONBOARDING_COMPLETE
+		) {
+			registry
+				.dispatch( CORE_UI )
+				.setValue(
+					UI_KEY_READER_REVENUE_MANAGER_SHOW_PUBLICATION_APPROVED_NOTIFICATION,
+					true
+				);
+		}
```
I think the idea here was to set this only if the onboarding state changed from something else to `ONBOARDING_COMPLETE`. With the current implementation, the [`PublicationApprovedOverlayNotification`](https://github.com/google/site-kit-wp/issues/8843) will be shown even if the user sets up RRM with a publication that had onboarding completed to start with.

@ankitrox Any inputs?

CC: @techanvil

techanvil replied:
Thanks for spotting this, @nfmohit, looks like we simply missed this one. I've created a followup issue to address it, https://github.com/google/site-kit-wp/issues/9089. I've left it with you in ACR, please take a look.

ankitrox replied:
Thanks for spotting this @nfmohit . I've created a follow-up PR for this issue https://github.com/google/site-kit-wp/pull/9090

techanvil replied:
Hey @ankitrox, thanks for raising the PR. However, with this issue now being closed, we'll need to create the fix against the newly created issue #9089.

ankitrox replied:
Thanks @techanvil 

I will keep that PR as is and we can link that PR to the issue when it is up for execution.

nfmohit replied:
Thank you for sorting this out, folks!
